### PR TITLE
Recommendation engine script

### PR DESCRIPTION
**Pull request by vansika**
script to generate HTML files containing essential information about training model and generating recommendations.

this script tries to follow conventions listed here: https://github.com/metabrainz/metabrainz-howto-guides/blob/master/writing_effective_data_processing_scripts.md

### CODE REVIEW

File: listenbrainz_spark/recommendations/create_dataframes.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    except Exception as err:
+        raise SystemExit("Cannot initialize Spark Session: %s. Aborting..." % (str(err)))
+
+    df = None
+    for y in range(config.STARTING_YEAR, config.ENDING_YEAR + 1):
+        for m in range(config.STARTING_MONTH, config.ENDING_MONTH + 1):
+            try:
+                month = listenbrainz_spark.sql_context.read.parquet('{}/data/listenbrainz/{}/{}.parquet'.format(config.HDFS_CLUSTER_URI, y, m))
+                df = df.union(month) if df else month
+            except Exception as err:
```
Comment (mayhem) [inline listenbrainz_spark/recommendations/create_dataframes.py]: Having this blanket exceptions is really not ideal. You should try and find exceptions that are appropriate for the given tools you are using. They seem to be poorly documented, but I cloned the spark repo and grepped in the pyspark dir and found:

  https://github.com/apache/spark/blob/master/python/pyspark/sql/utils.py

File: listenbrainz_spark/recommendations/create_dataframes.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        raise SystemExit("Parquet files containing listening history from {}-{} to {}-{} missing from HDFS".format(config.STARTING_YEAR, 
+                    "%02d" % config.STARTING_MONTH, config.ENDING_YEAR, "%02d" % config.ENDING_MONTH))
+    
+    print("\nRegistering Dataframe...")
+    table = 'df_to_train_{}'.format(datetime.strftime(datetime.utcnow(), '%Y_%m_%d'))
+    df.createOrReplaceTempView(table)
+    t = "%.2f" % (time.time() - ti)
+    print("Dataframe registered in %ss" % (t))
+
+    users_df = prepare_user_data(table)
```
Comment (mayhem) [inline listenbrainz_spark/recommendations/create_dataframes.py]: Here as well, we need to catch exceptions and report a proper error message.

File: listenbrainz_spark/recommendations/recommend.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+def recommend_user(user_name, model, all_recordings, recordings_df, count):
+    user_id = get_user_id(user_name)
+    user_playcounts = run_query("""
+        SELECT recording_id
+          FROM playcount
+         WHERE user_id = %d
+    """ % user_id)
+
+    candidate_recordings = all_recordings.subtract(user_playcounts)
```
Comment (mayhem) [inline listenbrainz_spark/recommendations/recommend.py]: As per our IRC conversation we should nuke this step.

File: listenbrainz_spark/recommendations/recommend.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        SELECT recording_id
+          FROM playcount
+         WHERE user_id = %d
+    """ % user_id)
+
+    candidate_recordings = all_recordings.subtract(user_playcounts) 
+    candidate_recordings.limit(count)
+    candidate_recordings = candidate_recordings.rdd.map(lambda r : r['recording_id'])
+
+    recommendations = model.predictAll(candidate_recordings.map(lambda recording: (user_id, recording))).takeOrdered(50, lambda product: -product.rating)
```
Comment (mayhem) [inline listenbrainz_spark/recommendations/recommend.py]: 50 should be defined in config or passed as a parameter.

File: listenbrainz_spark/recommendations/recommend.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+    path = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'recommendation-metadata.json')
+    with open(path, 'r') as f:
+        recommendation_metadata = json.load(f)
+        best_model_id = recommendation_metadata['best_model_id']
+
+    best_model_path = os.path.join('/', 'data', 'listenbrainz', 'recommendation-engine', 'best_model', '{}'.format(best_model_id))
+    
+    print("Loading model...")
+    for attempt in range(config.MAX_RETRIES):
```
Comment (mayhem) [inline listenbrainz_spark/recommendations/recommend.py]: Gah, this is ugly. (not your fault). I wonder if this gets better when we run on azure.

File: listenbrainz_spark/recommendations/train_models.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    validation_data.persist()
+    num_training = training_data.count()
+    num_validation = validation_data.count()
+    num_test = test_data.count()
+    print("Training model...")
+
+    for attempt in range(config.MAX_RETRIES):
+        try:
+            t0 = time.time()
+            model, model_metadata, best_model_metadata = train(training_data, validation_data, num_validation, [8, 12], [0.1, 10.0], [10, 20])
```
Comment (mayhem) [inline listenbrainz_spark/recommendations/train_models.py]: These values should also go into the config file.

File: listenbrainz_spark/recommendations/create_dataframes.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+def main():
+    ti = time.time()
+    try:
+        listenbrainz_spark.init_spark_session('Create_Dataframe')
+    except Exception as err:
+        raise SystemExit("Cannot initialize Spark Session: %s. Aborting..." % (str(err)))
+
+    df = None
+    for y in range(config.STARTING_YEAR, config.ENDING_YEAR + 1):
+        for m in range(config.STARTING_MONTH, config.ENDING_MONTH + 1):
```
[the remark this answers is not in the document]
Comment (vansika) [reply listenbrainz_spark/recommendations/create_dataframes.py]: Once we decide on how much back in time we will go to train the model for daily mix, we will create a function to calculate these values. Presently, they are in the config for testing purpose, for the ease of changing dates to any value.

File: listenbrainz_spark/recommendations/create_dataframes.py (hunk 1 above)
Comment (vansika) [reply listenbrainz_spark/recommendations/create_dataframes.py]: Thanks for the link:)

File: listenbrainz_spark/recommendations/recommend.py (hunk 5 above)
Comment (vansika) [reply listenbrainz_spark/recommendations/recommend.py]: What can be an alternative to this?

File: listenbrainz_spark/recommendations/create_dataframes.py [hunk 8]
```diff
@@ -111,64 +109,109 @@ def main():
...
     if df is None:
         raise SystemExit("Parquet files containing listening history from {}-{} to {}-{} missing from HDFS".format(config.STARTING_YEAR, 
                     "%02d" % config.STARTING_MONTH, config.ENDING_YEAR, "%02d" % config.ENDING_MONTH))
     
     print("\nRegistering Dataframe...")
     table = 'df_to_train_{}'.format(datetime.strftime(datetime.utcnow(), '%Y_%m_%d'))
-    df.createOrReplaceTempView(table)
+    try:
+        df.createOrReplaceTempView(table)
+    except Exception as err:
```
Comment (mayhem) [inline listenbrainz_spark/recommendations/create_dataframes.py]: Did you miss this one?

File: listenbrainz_spark/recommendations/recommend.py (hunk 5 above)
Comment (mayhem) [reply listenbrainz_spark/recommendations/recommend.py]: I suspect that this is a cause of poor performance of our VMs. An alternative is to rent better VMs.

**Pull request merged**